### ISSUE TITLE
[BUG] Function Wrapper yields unused-variable warning when compiled with pypy

### ISSUE THREAD

**Issue by WillAyd** [cython/cython#6099]
### Describe the bug

In pandas we have a function wrapper that looks like:

```python
def _make_error_func(func_name: str, cls):
    def f(*args, **kwargs):
        raise ValueError(f"NaTType does not support {func_name}")

    f.__name__ = func_name
    if isinstance(cls, str):
        # passed the literal docstring directly
        f.__doc__ = cls
    elif cls is not None:
        f.__doc__ = getattr(cls, func_name).__doc__
    return f
```

The generated c code looks something like:

```c
static struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func *__pyx_freelist_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func[8];
static int __pyx_freecount_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func = 0;

static PyObject *__pyx_tp_new_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func(PyTypeObject *t, CYTHON_UNUSED PyObject *a, CYTHON_UNUSED PyObject *k) {
  PyObject *o;
  #if CYTHON_COMPILING_IN_LIMITED_API
  allocfunc alloc_func = (allocfunc)PyType_GetSlot(t, Py_tp_alloc);
  o = alloc_func(t, 0);
  #else
  #if CYTHON_COMPILING_IN_CPYTHON
  if (likely((int)(__pyx_freecount_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func > 0) & (int)(t->tp_basicsize == sizeof(struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func)))) {
    o = (PyObject*)__pyx_freelist_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func[--__pyx_freecount_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func];
    memset(o, 0, sizeof(struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func));
    (void) PyObject_INIT(o, t);
  } else
  #endif
  {
    o = (*t->tp_alloc)(t, 0);
    if (unlikely(!o)) return 0;
  }
  #endif
  return o;
}

static void __pyx_tp_dealloc_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func(PyObject *o) {
  struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func *p = (struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func *)o;
  #if CYTHON_USE_TP_FINALIZE
  if (unlikely((PY_VERSION_HEX >= 0x03080000 || __Pyx_PyType_HasFeature(Py_TYPE(o), Py_TPFLAGS_HAVE_FINALIZE)) && __Pyx_PyObject_GetSlot(o, tp_finalize, destructor)) && (!PyType_IS_GC(Py_TYPE(o)) || !__Pyx_PyObject_GC_IsFinalized(o))) {
    if (__Pyx_PyObject_GetSlot(o, tp_dealloc, destructor) == __pyx_tp_dealloc_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func) {
      if (PyObject_CallFinalizerFromDealloc(o)) return;
    }
  }
  #endif
  Py_CLEAR(p->__pyx_v_func_name);
  #if CYTHON_COMPILING_IN_CPYTHON
  if (((int)(__pyx_freecount_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func < 8) & (int)(Py_TYPE(o)->tp_basicsize == sizeof(struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func)))) {
    __pyx_freelist_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func[__pyx_freecount_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func++] = ((struct __pyx_obj_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func *)o);
  } else
  #endif
  {
    #if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY
    (*Py_TYPE(o)->tp_free)(o);
    #else
    {
      freefunc tp_free = (freefunc)PyType_GetSlot(Py_TYPE(o), Py_tp_free);
      if (tp_free) tp_free(o);
    }
    #endif
  }
}
```

Compiling this with warnings yields `error: ‘__pyx_freecount_6pandas_5_libs_6tslibs_7nattype___pyx_scope_struct___make_error_func’ defined but not used [-Werror=unused-variable]`, which makes sense since the variable is only used when `CYTHON_COMPILING_IN_CPYTHON`

### Code to reproduce the behaviour:

See above

### Expected behaviour

No build warnings

### OS

Linux

### Python version

3.11

### Cython version

3.0.9

### Additional context

Example from pandas CI - https://github.com/pandas-dev/pandas/actions/runs/8348473859/job/22850478362?pr=57895#step:7:281

**Comment by matusvalo** [comment]
I think this is fixed (or at least changed) in master with following PR: https://github.com/cython/cython/pull/5885

**Comment by da-woods** [comment]
I'll backport the fix to the 3.0.x branch assuming it works.

**Comment by da-woods** [comment]
Closed in https://github.com/cython/cython/commit/d42e8776e50a2cf63ecaad8ee86eb5417cf2d7e2
